### PR TITLE
Feature/python benchmarking

### PR DESCRIPTION
**Pull request by vyasr**
This PR ports the benchmarks in https://github.com/vyasr/cudf_benchmarks, adding official benchmarks to the repository. The new benchmarks are designed from the ground up to make the best use of pytest, pytest-benchmark, and pytest-cases to simplify writing and maintaining benchmarks. Extended discussions of various previous design questions may be found on [the original repo](https://github.com/vyasr/cudf_benchmarks). Reviewers may also benefit from reviewing the companion PR creating documentation for how to write benchmarks, #11122.

Tests will not pass here until https://github.com/rapidsai/integration/pull/492 is merged.

### CODE REVIEW

**Comment by vyasr** [comment]
This PR has a large changeset by necessity. In order to make this review process a bit smoother, I have explicitly requested reviews from people who have contributed benchmarks before and/or are familiar with my design for the benchmarks (@isVoid, @shwina, and @galipremsagar).

File: python/cudf/benchmarks/conftest.py
Comment (isVoid) [inline python/cudf/benchmarks/conftest.py:92]: Let's put a comment here so that the assert error looks sensible when raised.

File: python/cudf/benchmarks/common/utils.py
Comment (isVoid) [inline python/cudf/benchmarks/common/utils.py:231]: Perhaps `make_fixture_and_inject_to_globals` is more descriptive?

File: python/cudf/benchmarks/common/utils.py
Comment (isVoid) [inline python/cudf/benchmarks/common/utils.py:254]: Why is sorted needed here?

File: python/cudf/benchmarks/common/utils.py
Comment (isVoid) [inline python/cudf/benchmarks/common/utils.py:29]: Should seed be a global configuration?

File: python/cudf/benchmarks/common/utils.py
Comment (isVoid) [inline python/cudf/benchmarks/common/utils.py:50]: It looks like this is valid syntax:
```python
In [5]: def flatten(xs):
   ...:     for x in xs:
   ...:         (yield x) if isinstance(x, str) else (yield from x)
```

**Review by isVoid — COMMENTED** [review]
Partial review part 1. Incredible work!

**Comment by ajschmidt8** [comment]
rerun tests

File: python/cudf/benchmarks/common/utils.py
Comment (vyasr) [reply python/cudf/benchmarks/common/utils.py:231]: It is more descriptive, but I think that's more detail than is really needed. This function is only two lines, and at that point the function name is almost as long as just copy-pasting the body of the function :) If someone really needs to understand the globals component they can read it.

File: python/cudf/benchmarks/common/utils.py
Comment (vyasr) [reply python/cudf/benchmarks/common/utils.py:29]: I guess the question is how would that global seed be used? Do we use the same seed for all functions? For example does `make_boolean_mask_column` also use the same seed?

File: python/cudf/benchmarks/common/utils.py
Comment (vyasr) [reply python/cudf/benchmarks/common/utils.py:50]: It is. More importantly though, `flatten` is no longer needed :upside_down_face: I changed the part of the code that was originally using it, and I can delete this altogether now.

File: python/cudf/benchmarks/common/config.py
Comment (vyasr) [reply python/cudf/benchmarks/common/config.py:13]: That is a very good point. `DEBUG_ONLY` seems like a reasonable choice.

**Comment by vyasr** [comment]
@ajschmidt8 no need to rerun tests here until we merge https://github.com/rapidsai/integration/pull/492

**Comment by ajschmidt8** [comment]
@vyasr, https://github.com/rapidsai/integration/pull/492 has been merged

**Review by ajschmidt8 — APPROVED** [review]
Approving `ops-codeowner` file changes

File: python/cudf/benchmarks/common/utils.py
Comment (isVoid) [inline python/cudf/benchmarks/common/utils.py:82]: It's not immediately clear to me what this means. Can you give an example?

File: python/cudf/benchmarks/common/utils.py
Comment (vyasr) [reply python/cudf/benchmarks/common/utils.py:187]: Nope. If you do
```
@accepts_cudf_benchmark(...)
@pytest.mark.parametrize(...)
def test_foo(...):
```
then marks are copied from `test_foo` onto the wrapper. If you swap that and do
```
@pytest.mark.parametrize(...)
@accepts_cudf_benchmark(...)
def test_foo(...):
```
then the marks are defined directly on the wrapper. In both cases, the wrapper gets all the marks.

File: python/cudf/benchmarks/common/utils.py
Comment (vyasr) [reply python/cudf/benchmarks/common/utils.py:108]: RangeIndex is not nullable and can't vary in dtype. In theory we could use the decorator for just sizes, I guess, but that seemed unnecessary. What do you think?

File: python/cudf/benchmarks/common/utils.py
Comment (vyasr) [reply python/cudf/benchmarks/common/utils.py:82]: Basically it means that you can write
```
@accepts_cudf_fixture(cls="column", dtype="int", nulls=False)
def test_foo(column)
```
and pytest will actually see a function
```
def test_foo(column_dtype_int_nulls_False)
``
Does that answer your question?

File: python/cudf/benchmarks/common/utils.py
Comment (isVoid) [reply python/cudf/benchmarks/common/utils.py:187]: Ah I see, didn't think of the other way.

File: python/cudf/benchmarks/common/utils.py
Comment (isVoid) [reply python/cudf/benchmarks/common/utils.py:82]: I guess my question is more with why `name` argument is needed for this utility function?

File: python/cudf/benchmarks/common/utils.py
Comment (bdice) [inline python/cudf/benchmarks/common/utils.py:24]: Why rounding instead of truncation, or requiring that inputs are `int` and not `Real`?

File: python/cudf/benchmarks/common/utils.py
Comment (bdice) [inline python/cudf/benchmarks/common/utils.py:111]:
```suggestion
        f"{', '.join(supported_classes)}"
```

File: python/cudf/benchmarks/common/utils.py
Comment (bdice) [inline python/cudf/benchmarks/common/utils.py:120]:
```suggestion
        f"{', '.join(column_generators)}"
```

File: python/cudf/benchmarks/common/utils.py
Comment (bdice) [inline python/cudf/benchmarks/common/utils.py:133]:
```suggestion
            f"only exist for the values {', '.join(NUM_COLS)}"
```

File: python/cudf/benchmarks/common/utils.py
Comment (bdice) [inline python/cudf/benchmarks/common/utils.py:141]:
```suggestion
            f"only exist for the values {', '.join(NUM_ROWS)}"
```

File: python/cudf/benchmarks/common/utils.py
Comment (bdice) [inline python/cudf/benchmarks/common/utils.py:148]:
```suggestion
        # pytest's test collection process relies on parsing the globals dict
```

File: python/cudf/benchmarks/common/utils.py
Comment (bdice) [inline python/cudf/benchmarks/common/utils.py:205]: Why not `*values`? Make it act more like `set`.
```suggestion
    def __init__(self, *values):
```

**Review by bdice — COMMENTED** [review]
This is great work. I have a lot of comments but I don't think any of them are major blockers. I'll hold off on approval until at least some of the suggestions are addressed, but I'd also be fine with merging this PR and fixing them later as well.

**Comment by vyasr** [comment]
@bdice I will be addressing your comments, but broadly speaking I'd like to avoid making any further changes to individual benchmarks in this PR. A number of your questions are suggestions for improvements to benchmarks themselves, and while those are super worthwhile I'm going to punt on those. I'm going to focus on the changes that affect the overall benchmarks framework (the fixture setup, the various helper functions, etc). But feel free to continue making those other suggestions, I'll leave those threads open and address them in the subsequent PRs.

File: python/cudf/benchmarks/common/utils.py
Comment (isVoid) [reply python/cudf/benchmarks/common/utils.py:108]: If I understand it correctly, including rangeindex here would change:
```python
def test_f(benchmark, rangeindex):
...
```
to
```python
@accept_cudf_fixture("rangeindex", size=[...])
def test_f(benchmark, rangeindex):
...
```
I don't think parameterizing the sizes is particularly useful for rangeindex as the default fixtures should suffice most use cases. I agree that we should leave it as a separate fixture union to avoid the complexity.

File: python/cudf/benchmarks/common/utils.py
Comment (isVoid) [reply python/cudf/benchmarks/common/utils.py:82]: I personally find it confusing to provide a way to alias your parameter, especially when the name of `cls` is unambiguous. The only usage of this is if we have very complex union in the future that even the name of `cls` becomes non-intelligible, then we can think of introducing the aliases.

I agree with your definition of the decorator in its current state. I think the debating point is whether we should take the decorator a step further, hiding more details from the developer. It's a good addition, but probably beyond what this PR has to offer.

Comment (vyasr) [reply python/cudf/benchmarks/common/utils.py:82]: Even if we have a very complex union the `cls` should never become unintelligible; that would only be possible if our class hierarchy become complex and we introduced weird classes like `IndexedFrameWithoutNulls` or something like that, which I don't think will ever happen. @shwina what do you think about removing support for the `name` parameter entirely?

Comment (vyasr) [reply python/cudf/benchmarks/common/utils.py:82]: > I agree with your definition of the decorator in its current state. I think the debating point is whether we should take the decorator a step further, hiding more details from the developer (with other decorators). It's a good addition, but probably beyond what this PR has to offer.

What do you mean by this? What other details would you want to hide? Are you suggesting something like `accepts_cudf_fixture_with_nulls`, i.e. putting more of the parameter information into decorators that wrap this one? I don't think that adds any value, but perhaps I'm misunderstanding your suggestion.

Comment (isVoid) [reply python/cudf/benchmarks/common/utils.py:82]: Apologies, you should read this paragraph in the context of the comments I posted in the other PR: https://github.com/rapidsai/cudf/pull/11122#discussion_r903866124, I meant specifically that the fixture that @shwina proposed is the advanced version that hides more detail.

Comment (vyasr) [reply python/cudf/benchmarks/common/utils.py:82]: OK, and in that scenario it's only "hiding" in the sense of the name of the decorator, right? The function signature is not changed at all.

File: python/cudf/benchmarks/common/utils.py
Comment (vyasr) [reply python/cudf/benchmarks/common/utils.py:108]: Yes, you understand correctly and we're in agreement that it doesn't seem very useful.

File: python/cudf/benchmarks/common/utils.py
Comment (vyasr) [reply python/cudf/benchmarks/common/utils.py:205]: That isn't how set works
```
>>> set(1, 2, 3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: set expected at most 1 argument, got 3
>>> set((1, 2, 3))
{1, 2, 3}
>>> set([1, 2, 3])
{1, 2, 3}
```

File: python/cudf/benchmarks/common/utils.py
Comment (vyasr) [reply python/cudf/benchmarks/common/utils.py:29]: I agree that this is mostly a concern for tests, not benchmarks, which is why I didn't think a whole lot about this yet. I think we can afford to punt on making this decision in this PR and then spend a bit of time investigating afterwards.

File: python/cudf/benchmarks/API/bench_dataframe.py
Comment (vyasr) [reply python/cudf/benchmarks/API/bench_dataframe.py:55]: The performance does matter for a couple of reasons, mostly around the fact that our sampling implementation is different for sampling rows vs columns and you actually _can't_ use cupy arrays for column sampling (because it needs to happen on the host) but you could use either for sampling rows and there is a meaningful perf difference.

File: python/cudf/benchmarks/common/utils.py
Comment (vyasr) [reply python/cudf/benchmarks/common/utils.py:24]: That's a question for @isVoid.

File: python/cudf/benchmarks/common/config.py
Comment (vyasr) [reply python/cudf/benchmarks/common/config.py:59]: You can do that, but what if we are using another plugin that also mucks with sys.path? I think the current solution of adding and then deleting is a minimally invasive solution that won't lead to breakage if another plugin also tries to do the same thing and doesn't include sufficient error checking.

File: python/cudf/benchmarks/common/utils.py
Comment (bdice) [reply python/cudf/benchmarks/common/utils.py:205]: 🤦‍♂️ Sorry, my bad.

File: python/cudf/benchmarks/common/config.py
Comment (vyasr) [reply python/cudf/benchmarks/common/config.py]: I converted the benchmarks to a package, so now we can use relative imports and this issue is moot

File: python/cudf/benchmarks/common/utils.py
Comment (vyasr) [reply python/cudf/benchmarks/common/utils.py]: Discussed offline, we are removing the name parameter but otherwise leaving the decorator as is (a single decorator with `cls` as a parameter rather than one decorator per `cls`).

**Comment by vyasr** [comment]
@bdice @isVoid I think that I have addressed or responded to all of the comments, aside from those focusing on specific benchmarks that I intend to push to future work.

File: ci/gpu/build.sh
Comment (bdice) [inline ci/gpu/build.sh:238]: I'd put this after normal pytests and before notebook tests -- unless there's a reason that we can fail earlier with good error messages by running this before standard pytests.

I would also add in the comment that this only runs a _tiny_ benchmark input and not a full set of benchmarks.

File: python/cudf/benchmarks/API/bench_multiindex.py
Comment (bdice) [inline python/cudf/benchmarks/API/bench_multiindex.py:3]: For consistency:
```suggestion
"""Benchmarks of MultiIndex methods."""

import numpy as np
```

File: python/cudf/benchmarks/internal/bench_column.py
Comment (bdice) [inline python/cudf/benchmarks/internal/bench_column.py:109]:
```suggestion
#           column (len(val) != len(key) and len == num_true)
```

**Review by bdice — APPROVED** [review]
Really awesome job @vyasr. Excited to have these! I left a couple small comments but I'm +1 for punting on anything you want to address later.

File: python/cudf/benchmarks/common/utils.py
Comment (isVoid) [reply python/cudf/benchmarks/common/utils.py:25]: I think requiring inputs are `int` is the way to go here. @vyasr I don't think you should address this in this PR. I'll make an issue after you merge this.

**Review by isVoid — APPROVED** [review]

File: python/cudf/benchmarks/API/bench_dataframe.py
Comment (galipremsagar) [inline python/cudf/benchmarks/API/bench_dataframe.py:11]: Use absolute import?

File: python/cudf/benchmarks/API/bench_frame_or_index.py
Comment (galipremsagar) [inline python/cudf/benchmarks/API/bench_frame_or_index.py:10]: Here too

File: python/cudf/benchmarks/API/bench_functions.py
Comment (galipremsagar) [inline python/cudf/benchmarks/API/bench_functions.py:8]: Here too

File: python/cudf/benchmarks/API/bench_functions_cases.py
Comment (galipremsagar) [inline python/cudf/benchmarks/API/bench_functions_cases.py:7]: Here too

File: python/cudf/benchmarks/API/bench_index.py
Comment (galipremsagar) [inline python/cudf/benchmarks/API/bench_index.py:8]: here too

File: python/cudf/benchmarks/API/bench_multiindex.py
Comment (galipremsagar) [inline python/cudf/benchmarks/API/bench_multiindex.py:9]: here

File: python/cudf/benchmarks/API/bench_series.py
Comment (galipremsagar) [inline python/cudf/benchmarks/API/bench_series.py:8]: here too

File: python/cudf/benchmarks/common/utils.py
Comment (galipremsagar) [inline python/cudf/benchmarks/common/utils.py:14]: Here

**Review by galipremsagar — COMMENTED** [review]
For consistency sake can we switch all the relative imports to absolute imports?

**Comment by vyasr** [comment]
> For consistency sake can we switch all the relative imports to absolute imports?

Done

**Review by galipremsagar — APPROVED** [review]

**Review by mroeschke — APPROVED** [review]
Looks pretty good! This is a good base of benchmarks that can be expanded on in the future.

Not sure if this was discussed, but how did you imagine these benchmarks to be used? For example in pandas, we have used ASV to catch performance regressions across verions, but we're inconsistent about this as it takes a while to run the whole suite.

**Comment by vyasr** [comment]
rerun tests

**Comment by vyasr** [comment]
> Looks pretty good! This is a good base of benchmarks that can be expanded on in the future.
> 
> Not sure if this was discussed, but how did you imagine these benchmarks to be used? For example in pandas, we have used ASV to catch performance regressions across verions, but we're inconsistent about this as it takes a while to run the whole suite.

We have indeed discussed this before. ASV is something that we've definitely considered, not for running benchmarks but rather using the dashboards with data from pytest-benchmark via some plugins written by other RAPIDS team members. For the moment, the process of tracking historical benchmark results is pretty manual. We've discussed automating benchmark runs more going forward, but we haven't come to any specific solution yet. I think that to some extent this is waiting on various changes from our ops team.

**Comment by vyasr** [comment]
rerun tests

File: ci/gpu/build.sh
Comment (vyasr) [inline ci/gpu/build.sh:251]:
```suggestion
gpuci_logger "Python pytest for cuDF benchmarks"
gpuci_logger "The path is ${PYTHONPATH}"
```

File: ci/gpu/build.sh
Comment (vyasr) [inline ci/gpu/build.sh:253]:
```suggestion
CUDF_BENCHMARKS_DEBUG_ONLY=ON py.test -n 8 python/cudf/benchmarks
CUDF_BENCHMARKS_USE_PANDAS=ON CUDF_BENCHMARKS_DEBUG_ONLY=ON py.test -n 8 python/cudf/benchmarks
```

**Comment by vyasr** [comment]
@bdice I had to roll back the changes making the benchmarks a Python package because it appears to change pytest's package discovery logic in a way that leads to it always finding the local copy of the package rather than the installed one, which is problematic when the package is not built in place (the case in our CI). I'm happy to revisit this further in the future, but in the interest of moving us forward I would like to merge as is and try to improve the organization and address the CI issues in future PRs.

**Pull request merged**
